File: app/src/main/java/com/test/gifimageview/GIFImageView.java
package com.test.gifimageview;

import android.content.Context;
import android.util.AttributeSet;
import android.util.Log;

import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatImageView;

/**
 * @Author: duke
 * @DateTime: 2021-10-15 18:14:06
 * @Description: 功能描述：
 */
public class GIFImageView extends AppCompatImageView {

    public GIFImageView(Context context) {
        this(context, null);
    }

    public GIFImageView(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public GIFImageView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context, attrs, defStyleAttr);
    }

    private void init(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        Log.e("", "");
    }

}

File: app/src/main/java/com/test/gifimageview/GIFImageView.kt
package com.test.gifimageview

import android.content.Context
import android.graphics.Canvas
import android.graphics.Movie
import android.graphics.drawable.AnimatedImageDrawable
import android.util.AttributeSet
import androidx.appcompat.widget.AppCompatImageView

/**
 * @Author: duke
 * @DateTime: 2021-10-15 18:14:06
 * @Description: 功能描述：Gif ImageView
 */
class GIFImageView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : AppCompatImageView(context, attrs, defStyleAttr) {

    private var gifDrawable: AnimatedImageDrawable? = null
    private var autoGIFSize: Boolean = false

    fun setGIFFromAssets(assetFileName: String) {
        if (android.os.Build.VERSION.SDK_INT < android.os.Build.VERSION_CODES.P) {
            return
        }
        val stream = resources?.assets?.open(assetFileName)
        val tempDrawable = AnimatedImageDrawable.createFromStream(stream, null)
        gifDrawable = (tempDrawable as? AnimatedImageDrawable?)
        if (gifDrawable == null) {
            return
        }
        gifDrawable?.start()
        invalidate()
    }

    fun setAutoGIFSize(isAuto: Boolean) {
        autoGIFSize = isAuto
        requestLayout()
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)

        if (gifDrawable == null) {
            return
        }
        val gifWidth = gifDrawable?.intrinsicWidth ?: 0
        val gifHeight = gifDrawable?.intrinsicHeight ?: 0
        if (gifWidth == 0 || gifHeight == 0) {
            return
        }
        if (!autoGIFSize) {
            return
        }
        val widthSpec = MeasureSpec.makeMeasureSpec(gifWidth, MeasureSpec.EXACTLY)
        val heightSpec = MeasureSpec.makeMeasureSpec(gifHeight, MeasureSpec.EXACTLY)
        setMeasuredDimension(widthSpec, heightSpec)
    }

    override fun onDraw(canvas: Canvas?) {
        super.onDraw(canvas)
        playGIF(canvas)
    }

    private fun playGIF(canvas: Canvas?) {
        if (canvas == null || gifDrawable == null) {
            return
        }
        gifDrawable?.draw(canvas)
    }

}
